spanner/grpc_client.go

```diff
@@ -171,7 +174,64 @@ func (g *grpcSpannerClient) DeleteSession(ctx context.Context, req *spannerpb.De
 	return err
 }
 
+// setSpanAttributes dynamically sets span attributes based on the request type.
+func setSpanAttributes[T any](span oteltrace.Span, req T) {
+	if !span.IsRecording() {
+		return
+	}
+
+	var attrs []attribute.KeyValue
+
+	if r, ok := any(req).(interface {
+		GetRequestOptions() *spannerpb.RequestOptions
+	}); ok {
+		if tag := r.GetRequestOptions().GetTransactionTag(); tag != "" {
+			attrs = append(attrs, attribute.String("transaction.tag", tag))
+		}
+		if tag := r.GetRequestOptions().GetRequestTag(); tag != "" {
+			attrs = append(attrs, attribute.String("statement.tag", tag))
+		}
+	}
+
+	if r, ok := any(req).(interface{ GetSql() string }); ok {
+		if sql := r.GetSql(); sql != "" {
+			attrs = append(attrs, attribute.String("db.statement", sql))
+		}
+	} else if r, ok := any(req).(interface {
+		GetStatements() []*spannerpb.ExecuteBatchDmlRequest_Statement
+	}); ok {
+		if stmts := r.GetStatements(); len(stmts) > 0 {
+			sqls := make([]string, len(stmts))
+			for i, stmt := range stmts {
+				sqls[i] = stmt.GetSql()
+			}
+			attrs = append(attrs, attribute.StringSlice("db.statement", sqls))
+		}
+	}
+
+	if r, ok := any(req).(interface{ GetTable() string }); ok {
+		if table := r.GetTable(); table != "" {
+			attrs = append(attrs, attribute.String("db.table", table))
+		}
+	}
+
+	span.SetAttributes(attrs...)
+}
+
+func setGFEAndAFESpanAttributes(span oteltrace.Span, latencyMap map[string]time.Duration) {
+	if !span.IsRecording() {
+		return
+	}
+	for t, v := range latencyMap {
+		span.SetAttributes(
+			attribute.Float64(t[:3]+".latency_ms", float64(v.Nanoseconds())/1e6),
```

Comment:
nit: this could use a small comment explaining what `t[:3]` does (gfe or afe latency). And are we absolutely sure that this will never change in the future? What if we introduce some new header key in the future that is shorter, and someone is still using this version of the client? Maybe we should only do the `t[:3]` for the well-known header keys, and include the full header key for other values (or completely skip them).